---
title: "Widespread introgression across a phylogeny of 155 Drosophila genomes"
author: "Yilin Bai"
date: "2023-08-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## **Part 1: Introduction and Methodology**

**Introduction:**

The study addresses the ongoing debate surrounding the extent of gene exchange, known as introgression, in evolutionary processes. Genomic data has revealed the prevalence of introgression across diverse taxa, impacting evolutionary outcomes and adaptive processes. While introgression was initially considered detrimental, it is now recognized as a source of genetic variation driving local adaptation and diversification. The study aims to systematically analyze the frequency of introgression across the phylogenetic tree of the Drosophila genus, leveraging clade-level genomic data and without biased selection of hybridizing taxa.

**Method:**

The researchers analyze 155 Drosophila genomes and 4 outgroup genomes to construct a robust phylogeny. They utilize genome-scale sequence data to infer phylogenetic relationships, employing both maximum-likelihood and coalescent-based methods. The inclusion of fossil calibrations ensures accurate estimation of divergence times. The phylogenetic analyses yield high-confidence relationships among most species, with minor discrepancies noted for a few nodes. Additionally, the study employs protein supermatrix analysis and Quartet Sampling to validate the inferred species tree topology and assess discordance patterns that might indicate introgression.



## **Part 2: Data and Analysis**

**Dataset:**

The dataset encompasses 155 Drosophila genomes and 4 outgroup genomes, representing a broad evolutionary timespan. The researchers annotate and align sequence data for Benchmarking Universal Single-Copy Orthologs (BUSCOs), totaling over 8 million nucleotide positions. These genomes cover over 50 million years of Drosophila evolution and include multiple samples from 9 major radiations within the genus. Fossil calibrations, including those from Drosophilidae fossils and geological information, contribute to the precise estimation of divergence times.



## **Part 3: Results**

a.  **Tree Topology:**

    Include a figure that displays the different tree topologies. You can use a tree visualization tool or software (e.g., FigTree) to create rooted trees with branch lengths. This will allow readers to visually compare the differences in tree structure.
    
    ```{r}
# Command of run one class model iq-tree
 iqtree_one_model_command <- '% /Users/lindsaybai/Desktop/BIOL8706/iqtree-2.2.2.7.modelmix-MacOSX/bin/iqtree2 -s /Users/lindsaybai/Desktop/BIOL8706/one_gene/nad1alignment.fasta   -B 1000 -T AUTO'

# Command of run mixture class model iq-tree
iqtree_mixture_command <- '/Users/lindsaybai/Desktop/BIOL8706/iqtree-2.2.2.7.modelmix-MacOSX/bin/iqtree2  -s /Users/lindsaybai/Desktop/BIOL8706/one_gene_mixture/nad1alignment.fasta   -m "ESTMIXNUM" -mrate E,I,G,I+G,R,I+R -opt_qmix_criteria 1'


system(iqtree_one_model_command)
system(iqtree_mixture_command)
```


    ```{r}
    library(Biostrings)
    # Replace 'sequences.fasta' with your actual file path
    sequences <- readDNAStringSet('~/Desktop/BIOL8706/phylogenetic_project/Phylogenetic_project/Data/mixture_class_single_gene/EOG09150A0E_extraction.fna')
    sequences
    ```

    #### Plot trees side by side

    ```{r}
    library(ape)
    library(phytools)
    # Read the tree files
    mix_tree <- read.tree("~/Desktop/BIOL8706/phylogenetic_project/Phylogenetic_project/Data/one_class_single_gene/one_tree.treefile")
    one_tree <- read.tree("~/Desktop/BIOL8706/phylogenetic_project/Phylogenetic_project/Data/mixture_class_single_gene/mix_tree.treefile")
    ```

    ```{r fig.height=6, fig.width=14}
    par(mfrow = c(1, 2))  # Set the plotting layout to 1 row and 2 columns
    plot(one_tree, main = "One class model tree")  # Plot the first tree
    plot(mix_tree, main = "Mixture class model tree")  # Plot the second tree
    ```

    #### **Cophylogeny plot**

    ```{r}
    ## create co-phylogenetic object
    wasp.cophylo<-cophylo(mix_tree,one_tree)
    ## plot co-phylogenies
    plot(wasp.cophylo,link.type="curved",link.lwd=4,
     link.lty="solid",link.col=make.transparent("red",
     0.25))
    par(mar=c(5.1,4.1,4.1,2.1))
    ```

b.  **Branch Lengths**

    ```{r}
branch_lengths_equal <- all.equal.phylo(mix_tree, one_tree)
cat("Branch lengths are equal:", is.null(branch_lengths_equal), "\n")
```

    ```{r}

library(phangorn)
# Calculate the Robinson-Foulds distance
rf_distance <- treedist(one_tree, mix_tree)

# Display the RF distance
cat("Robinson-Foulds distance:", rf_distance, "\n")
```

c.  **Best Model and Parameters:**

    ### Read the IQ-TREE file

    ```{r}
one_line <- readLines('~/Desktop/BIOL8706/phylogenetic_project/Phylogenetic_project/Data/one_class_single_gene/EOG09150A0E_extraction.fna.log')
mix_line <- readLines("~/Desktop/BIOL8706/phylogenetic_project/Phylogenetic_project/Data/mixture_class_single_gene/EOG09150A0E_extraction.fna.log")
```

    ### Extract best-fit model(one-class & mixture class)

    ```{r}
one_model <- grep("Best-fit model:", one_line, value = TRUE)
mix_model <- grep("^  Best-fit Q-Mixture model:", mix_line, value = TRUE)
one_best_model <- sub("Best-fit model: ([^\\s]+)\\s.*", "\\1", one_model)
mix_best_model <- sub("^  Best-fit Q-Mixture model: (.*)", "\\1", mix_model)
```
.

d.  **BIC Score, AIC Score, and Likelihood:**

    ### Exact best-fit model parameters(one-class & mixture class)

    ```{r}
# Find indices of lines starting with 'Rate parameters'
rate_params_indices_1 <- grep("^Rate parameters:", one_line)
rate_params_indices_2 <- grep("^Rate parameters:", mix_line)

# Extract the last line with rate parameters
last_rate_params_line_1 <- sub("^Rate parameters: (.*)", "\\1",one_line[rate_params_indices_1[length(rate_params_indices_1)]])

last_rate_params_line_2 <- sub("^Rate parameters: (.*)", "\\1",mix_model[rate_params_indices_2[length(rate_params_indices_2)]])

```

    ### Exact best-fit model BIC score(one-class & mixture class)

    ```{r warning=FALSE}
# Escape special characters in the best model name
escaped_best_model <- gsub("([.+])", "\\\\\\1", one_best_model)

# Search for the line containing the best model
bic_line <- grep(escaped_best_model, one_line, value = TRUE)

# Extract BIC score from the bic_line
bic_score <- na.omit(as.numeric(sub(".*\\s(\\d+\\.\\d+)\\s*$", "\\1", bic_line)))
```


e.  **Discussion:**

    ### Display extracted information

```{r}

# Create a data frame
output_data <- data.frame(
  "One class model" = c(one_best_model, last_rate_params_line_1, bic_score),
  "Mixture class model" = c(mix_best_model, last_rate_params_line_2, ""),
  row.names = c("Best Model:", "Parameters:", "BIC Score:")
)

# Print the formatted table
print(output_data)
```

f.  **Conclusion:**

    Summarize the key findings of the analysis, discussing any insights gained from comparing the different aspects of the phylogenetic analysis.

g.  **References:**


    Suvorov A, Kim BY, Wang J, Armstrong EE, Peede D, D'Agostino ERR, Price DK, Waddell P, Lang M, Courtier-Orgogozo V, David JR, Petrov D, Matute DR, Schrider DR, Comeault AA. Widespread introgression across a phylogeny of 155 Drosophila genomes. Curr Biol. 2022 Jan 10;32(1):111-123.e5. doi: 10.1016/j.cub.2021.10.052. Epub 2021 Nov 16. PMID: 34788634; PMCID: PMC8752469.

    Suvorov, Anton; Kim, Bernard; Wang, Jeremy; Armstrong, Ellie; Peede, David; D'Agostino, Emmanuel R. R.; et al. (2020). Widespread introgression across a phylogeny of 155 Drosophila genomes. figshare. Dataset. <https://doi.org/10.6084/m9.figshare.13264697.v1>
