---
title: "Sergestidae shrimp single-gene alignment"
author: "Yilin Bai"
date: "2023-08-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,message = FALSE, warning = FALSE)
```

## Paper and dataset

The data set I chose contains Sergestidae shrimp family single- and multi-gene alignments and tree files.


### **1. Paper Reference:**

Golightly, C., DeLeo, D. M., Perez, N., Chan, T.-Y., Landeira, J. M., & Bracken‚ÄêGrissom, H. D. (2022). Tracing the evolution of bioluminescent light organs across the deep-sea shrimp family Sergestidae using a genomic skimming and phylogenetic approach. Invertebrate Systematics, 36(1), 22-35. https://doi.org/10.1071/IS21013


### 2. **Dataset Reference:**

Golightly, Charles; Bracken-Grissom, Heather (2022), Sergestidae shrimp family single- and multi-gene alignments and tree files, Dryad, Dataset, https://doi.org/10.5061/dryad.vdncjsxvk

## Dataset discription

The provided information describes a comprehensive database centered around the investigation of the evolution of bioluminescence in deep-sea shrimp of the Sergestidae family. The database encompasses various aspects of the research, including its focus, methods, and approaches. Here's a more detailed overview of the database itself:


**Data Collection and Sample Acquisition:**
The database provides insights into the collection process of the specimens. The deep-sea shrimp are collected through mid-water trawls in specific geographic regions, notably the Gulf of Mexico and Florida Straits. This data collection process involves capturing live organisms from their natural habitats, ensuring that the samples represent the biological diversity of the family Sergestidae.

**Targeted Genes and Techniques:**
The genetic investigation involves the amplification and sequencing of specific genes. The targeted genes include H3, NaK, COI, and 16S. These genes are of particular interest due to their relevance to the study's focus on bioluminescence evolution. PCR protocols are employed to amplify these genes for sequencing, and Sanger sequencing techniques are used for their analysis.

**Advanced Techniques - Genome Skimming:**
The database introduces an advanced technique known as "genome skimming." This approach involves the fragmentation of extracted genomic DNA and the subsequent preparation of libraries using NEB Next Ultra II Library Prep kits. High-throughput sequencing using Illumina HiSeq 3000/4000 technology is employed to capture mitochondrial genomic data from the sampled individuals.

**Bioinformatics and Analysis:**
The database explains the subsequent bioinformatics steps involved in the analysis of the generated genetic data. Tools such as NOVOplasty are utilized for the assembly of complete mitochondrial genomes. Annotation of these genomes is performed using the MITOS online web tool. Bioinformatics software, notably Geneious Prime, plays a pivotal role in handling, cleaning, and organizing the raw sequence data for further analysis.

**Phylogenetic Analysis and Ancestral State Reconstruction:**
The core analytical components of the study involve constructing phylogenetic trees using the generated genetic data. The database provides insights into the methodologies employed for alignment creation using MAFFT and tree construction using IQTREE and MrBayes. Additionally, ancestral state reconstruction techniques are used to infer the ancestral states of bioluminescent traits in different lineages.


## Visualization of alignment 



## Choose a gene


```{r}
library(seqinr)
fasta_data <- read.fasta("nad1alignment.fasta")
```


## Run IQ-TREE 
```{r}

# Command of run one class model iq-tree
 iqtree_one_model_command <- '% /Users/lindsaybai/Desktop/BIOL8706/iqtree-2.2.2.7.modelmix-MacOSX/bin/iqtree2 -s /Users/lindsaybai/Desktop/BIOL8706/one_gene/nad1alignment.fasta   -B 1000 -T AUTO'

# Command of run mixture class model iq-tree
iqtree_mixture_command <- '/Users/lindsaybai/Desktop/BIOL8706/iqtree-2.2.2.7.modelmix-MacOSX/bin/iqtree2  -s /Users/lindsaybai/Desktop/BIOL8706/one_gene_mixture/nad1alignment.fasta   -m "ESTMIXNUM" -mrate E,I,G,I+G,R,I+R -opt_qmix_criteria 1'


system(iqtree_one_model_command)
system(iqtree_mixture_command)
```


## Compare the output
### Read trees from files
```{r}
library(ape)
tree1 <- read.tree("nad1alignment_onemodle.fasta.treefile")
tree2 <- read.tree("nad1alignment_mixture.fasta.treefile")

```

### Read the IQ-TREE file
```{r}
log_contents_onegene <- readLines("nad1alignment_onegene.fasta.log")
log_contents_mixture <- readLines("nad1alignment_mixture.fasta.log")
```


### Extract best-fit model(one-class & mixture class)

```{r}
best_model_line_1 <- grep("Best-fit model:", log_contents_onegene, value = TRUE)
best_model_line_2 <- grep("^  Best-fit Q-Mixture model:", log_contents_mixture, value = TRUE)
tree1_best_model <- sub("Best-fit model: ([^\\s]+)\\s.*", "\\1", best_model_line_1)
tree2_best_model <- sub("^  Best-fit Q-Mixture model: (.*)", "\\1", best_model_line_2)
```


### Exact best-fit model parameters(one-class & mixture class)

```{r}
# Find indices of lines starting with 'Rate parameters'
rate_params_indices_1 <- grep("^Rate parameters:", log_contents_onegene)
rate_params_indices_2 <- grep("^Rate parameters:", log_contents_mixture)

# Extract the last line with rate parameters
last_rate_params_line_1 <- log_contents_onegene[rate_params_indices_1[length(rate_params_indices_1)]]
last_rate_params_line_2 <- log_contents_onegene[rate_params_indices_2[length(rate_params_indices_2)]]
```

### Exact best-fit model BIC score(one-class & mixture class)

```{r warning=FALSE}
# Escape special characters in the best model name
escaped_best_model <- gsub("([.+])", "\\\\\\1", tree1_best_model)

# Search for the line containing the best model
bic_line <- grep(escaped_best_model, log_contents_onegene, value = TRUE)

# Extract BIC score from the bic_line
bic_score <- na.omit(as.numeric(sub(".*\\s(\\d+\\.\\d+)\\s*$", "\\1", bic_line)))
```


### Display extracted information

```{r}
cat("Best one class model:", tree1_best_model, "\n")
cat("Best mixture model:", tree2_best_model, "\n")

cat("One class model", last_rate_params_line_1, "\n")
cat("Mixture model", last_rate_params_line_2, "\n")

cat("One class model BIC score:", bic_score, "\n")

```


### Compare tree topologies

```{r}
library(ape)
topology_equal <- identical(tree1$edge, tree2$edge)
cat("Tree topologies are equal:", topology_equal, "\n")
topology_distance <- dist.topo(tree1, tree2)
topology_distance
```


### Compare branch lengths

```{r}
branch_lengths_equal <- all.equal.phylo(tree1, tree2)
cat("Branch lengths are equal:", is.null(branch_lengths_equal), "\n")
```


```{r}

library(phangorn)
# Calculate the Robinson-Foulds distance
rf_distance <- treedist(tree1, tree2)

# Display the RF distance
cat("Robinson-Foulds distance:", rf_distance, "\n")
```

### Visual Comparison:

#### Plot trees side by side 
```{r}
par(mfrow = c(1, 2))  # Set the plotting layout to 1 row and 2 columns
plot(tree1, main = "Tree 1")  # Plot the first tree
plot(tree2, main = "Tree 2")  # Plot the second tree
```


#### Create the cophylogeny plot
```{r}
cophyloplot(tree1, tree2)
```


### Conclusion